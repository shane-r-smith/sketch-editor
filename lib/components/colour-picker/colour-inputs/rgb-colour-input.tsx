import TextField from "@mui/material/TextField";
import type { BaseColourPickerProps } from "../colour-picker.types";
import { colord } from "colord";
import { useCallback, useEffect, useId, useState } from "react";
import { stringToClampedRgb } from "../colour-picker.utils";
import Grid from "@mui/material/Grid";
import FormLabel from "@mui/material/FormLabel";
import { Stack } from "@mui/material";

export function RgbColourInput({
  colour,
  type,
  onChange,
  ...rest
}: BaseColourPickerProps) {
  const [currentColour, setCurrentColour] = useState(colord(colour).toRgb());

  const redId = useId();
  const greenId = useId();
  const blueId = useId();

  const handleChange = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) => {
      try {
        const isRed = event.target.id === redId;
        const isGreen = event.target.id === greenId;
        const isBlue = event.target.id === blueId;

        const newValue = stringToClampedRgb(event.target.value);

        const updatedColour = structuredClone(currentColour);

        if (isRed) {
          updatedColour.r = newValue;
        }

        if (isGreen) {
          updatedColour.g = newValue;
        }

        if (isBlue) {
          updatedColour.b = newValue;
        }

        setCurrentColour(updatedColour);
        onChange(colord(updatedColour).toRgbString(), type);
      } catch {
        // adsorb any errors
      }
    },
    [blueId, currentColour, greenId, onChange, redId, type]
  );

  useEffect(() => {
    try {
      const updatedColour = colord(currentColour).toRgbString();
      if (updatedColour === colord(colour).toRgbString()) {
        return;
      }

      const value = colord(colour);

      if (value.isValid()) {
        setCurrentColour(value.toRgb());
        return;
      }

      setCurrentColour({ r: 0, g: 0, b: 0, a: 1 });
    } catch {
      // adsorb any errors
    }
  }, [colour, currentColour, onChange, type]);

  return (
    <Grid container spacing={0.5} columns={3}>
      <Stack>
        <FormLabel className="absolute -top-6 text-sm font-medium">R</FormLabel>
        <TextField
          {...rest}
          id={redId}
          value={currentColour.r}
          onChange={handleChange}
          slotProps={{
            htmlInput: {
              type: "number",
              min: 0,
              max: 255,
            },
          }}
          name="red"
          size="small"
        />
      </Stack>
      <Stack>
        <FormLabel className="absolute -top-5">G</FormLabel>
        <TextField
          {...rest}
          id={greenId}
          value={currentColour.g}
          onChange={handleChange}
          slotProps={{
            htmlInput: {
              type: "number",
              min: 0,
              max: 255,
            },
          }}
          name="green"
          size="small"
        />
      </Stack>
      <Stack>
        <FormLabel className="absolute -top-5">B</FormLabel>
        <TextField
          {...rest}
          id={blueId}
          value={currentColour.b}
          onChange={handleChange}
          slotProps={{
            htmlInput: {
              type: "number",
              min: 0,
              max: 255,
            },
          }}
          name="blue"
          size="small"
        />
      </Stack>
    </Grid>
  );
}
