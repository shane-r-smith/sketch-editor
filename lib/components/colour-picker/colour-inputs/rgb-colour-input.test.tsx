import { fireEvent, render } from "@testing-library/react";
import type { BaseColourPickerProps } from "../colour-picker.types";
import { RgbColourInput } from "./rgb-colour-input";

describe("<RgbColourInput />", () => {
  const defaultProps: BaseColourPickerProps = {
    type: "RGB",
    colour: "rgb(255, 0, 0)",
    onChange: vi.fn(),
  };

  it("renders correctly and matches snapshot", () => {
    const { container } = render(<RgbColourInput {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("converts and renders correctly and matches snapshot when provided with HEX colour", () => {
    // Arrange
    // Act
    const { container } = render(
      <RgbColourInput {...defaultProps} colour="#0000FF" />
    );

    // Assert
    const inputs = container.querySelectorAll("input")!;
    expect(inputs).toMatchSnapshot();
  });

  it("calls onChange with updated value and type red input changes", async () => {
    const { container } = render(<RgbColourInput {...defaultProps} />);

    const input = container.querySelector("input[name='red']")!;

    fireEvent.change(input, { target: { value: "0" } });

    expect(defaultProps.onChange).toHaveBeenCalledWith("rgb(0, 0, 0)", "RGB");
  });

  it("calls onChange with updated value and type green input changes", async () => {
    const { container } = render(<RgbColourInput {...defaultProps} />);

    const input = container.querySelector("input[name='green']")!;

    fireEvent.change(input, { target: { value: "255" } });

    expect(defaultProps.onChange).toHaveBeenCalledWith(
      "rgb(255, 255, 0)",
      "RGB"
    );
  });

  it("calls onChange with updated value and type green input changes", async () => {
    const { container } = render(<RgbColourInput {...defaultProps} />);

    const input = container.querySelector("input[name='blue']")!;

    fireEvent.change(input, { target: { value: "255" } });

    expect(defaultProps.onChange).toHaveBeenCalledWith(
      "rgb(255, 0, 255)",
      "RGB"
    );
  });

  it("handles invalid colour input gracefully", async () => {
    const { container } = render(<RgbColourInput {...defaultProps} />);

    const input = container.querySelector("input[name='red']")!;
    fireEvent.change(input, { target: { value: "invalid" } });

    expect(defaultProps.onChange).toHaveBeenCalledWith("rgb(0, 0, 0)", "RGB");
  });
});
