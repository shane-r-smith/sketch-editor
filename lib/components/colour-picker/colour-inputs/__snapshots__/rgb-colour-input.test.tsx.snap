// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<RgbColourInput /> > converts and renders correctly and matches snapshot when provided with HEX colour 1`] = `
NodeList [
  <input
    aria-invalid="false"
    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
    id="«r6»"
    max="255"
    min="0"
    name="red"
    type="number"
    value="0"
  />,
  <input
    aria-invalid="false"
    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
    id="«r7»"
    max="255"
    min="0"
    name="green"
    type="number"
    value="0"
  />,
  <input
    aria-invalid="false"
    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
    id="«r8»"
    max="255"
    min="0"
    name="blue"
    type="number"
    value="255"
  />,
]
`;

exports[`<RgbColourInput /> > renders correctly and matches snapshot 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-row MuiGrid-spacing-xs-0.5 css-1d2jkxl-MuiGrid-root"
  >
    <div
      class="MuiStack-root css-nen11g-MuiStack-root"
    >
      <label
        class="MuiFormLabel-root MuiFormLabel-colorPrimary absolute -top-6 text-sm font-medium css-nwsdb5-MuiFormLabel-root"
      >
        R
      </label>
      <div
        class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
            id="«r0»"
            max="255"
            min="0"
            name="red"
            type="number"
            value="255"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1nf2c5d-MuiNotchedOutlined-root"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiStack-root css-nen11g-MuiStack-root"
    >
      <label
        class="MuiFormLabel-root MuiFormLabel-colorPrimary absolute -top-5 css-nwsdb5-MuiFormLabel-root"
      >
        G
      </label>
      <div
        class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
            id="«r1»"
            max="255"
            min="0"
            name="green"
            type="number"
            value="0"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1nf2c5d-MuiNotchedOutlined-root"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiStack-root css-nen11g-MuiStack-root"
    >
      <label
        class="MuiFormLabel-root MuiFormLabel-colorPrimary absolute -top-5 css-nwsdb5-MuiFormLabel-root"
      >
        B
      </label>
      <div
        class="MuiFormControl-root MuiTextField-root css-1xp5r68-MuiFormControl-root-MuiTextField-root"
      >
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-quhxjy-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1pzfmz2-MuiInputBase-input-MuiOutlinedInput-input"
            id="«r2»"
            max="255"
            min="0"
            name="blue"
            type="number"
            value="0"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1nf2c5d-MuiNotchedOutlined-root"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>
</div>
`;
